### PR TITLE
Closes #11

### LINKED ISSUE #11: Integrate Aloha with H2O
Add the ability to build H2O Aloha models.  This should have all the same support that VW currently has.

### PR DESCRIPTION
**Pull request by deaktator**
I still need to add some tests but the code seems pretty robust. Tests obviously might flesh out edge cases that need to be hardened.

Still need to add implicit conversion functions to `com.eharmony.aloha.feature.BasicFunctions` in `aloha-core` to make the grammar nicer to work with. _E.g._ not using `Option` in the feature specifications.

### CODE REVIEW

File: aloha-h2o/pom.xml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    <distributionManagement>
+        <site>
+            <id>${site.deploy.id}</id>
+            <!-- Note that GitHub does not like having leading slash here -->
+            <url>${site.deploy.url}${project.artifactId}</url>
+        </site>
+    </distributionManagement>
+
+    <properties>
```
Comment (jmorra) [inline aloha-h2o/pom.xml:25]: This version seems a little old.  Do you think it's worthwhile to upgrade to 3.6.0.3 as outline [here](http://h2o.ai/download/).  Also, if I remember correctly H2O guarantees NO cross version compatibility right?

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  @transient private[this] lazy val lazyAnyRefFF = featureFunctions map {
+    case DoubleFeatureFunction(f) => f
+    case StringFeatureFunction(f) => f
+  }
+
+  @transient private[this] lazy val h2oPredictor: RowData => Either[IllConditioned, B] = {
+    val sourceFile = new java.io.File(modelSource.localVfs.descriptor)
+    val p = getH2oPredictor(sourceFile, _.fromFile).get
+    if (modelSource.shouldDelete)
+      Try[Unit] { sourceFile.delete() }
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala:56]: What is this syntax for?

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/categories/package.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    TypeCoercion[C, B] map { coercion =>
+      predictionFn(model) andThen (_.right map coercion)
+    } toRight noCoercionFound(model)
+
+  private[this] def noCoercionFound(model: EasyPredictModelWrapper) =
+    TypeCoercionNotFound(model.getModelCategory)
+}
+
+object H2oModelCategory {
+  private[categories]def wellConditioned(d: Double) = NaN              != d &&
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/categories/package.scala]: Does equals work here?  I know in Java NaN != NaN.

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+extends AlohaReadable[Try[B]]
+   with ReadableByString[Try[B]] {
+
+  /** Read from a String.
+    * @param code a String to read.
+    * @return the result
+    */
+  override def fromString(code: String): Try[B] = {
+    val diagnosticCollector = new DiagnosticCollector[JavaFileObject]
+
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala:41]: What is this?

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    } yield instance
+  }
+
+  def getCompileDir(dir: Option[File]) = dir.map(d => Try{(d, false)}) getOrElse Compiler.tmpDir.map(d => (d, true))
+
+  def getCompileTask(compilationUnit: InMemoryJavaSource[B],
+                     compileDir: File,
+                     diagnosticCollector: DiagnosticCollector[JavaFileObject]) = Try {
+    val locations = asJavaIterable(Seq("-d", compileDir.getCanonicalPath))
+    val compiler = ToolProvider.getSystemJavaCompiler
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala:62]: This would seem to imply that javac must be on the machine running the model.  Is this so?

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala [hunk 6]
```diff
@@ -0,0 +1,116 @@
+package com.eharmony.aloha.models.h2o.compiler
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala:1]: Is this code you snagged from somewhere else?  If so could you provide the link.

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/illConditioned.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.eharmony.aloha.models.h2o
+
+/**
+ * H2o likes to return `NaN` as a "''valid''" prediction, especially when covariate data is missing.
+ * This is not really acceptable.  So, we actually
```
Comment (jmorra) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/illConditioned.scala:5]: Finish this sentence...

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala (hunk 5 above)
Comment (deaktator) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/compiler/compiler.scala:62]: Not necessarily javac but definitely (jar files in) JDK, if that makes sense.

File: aloha-h2o/pom.xml (hunk 1 above)
Comment (deaktator) [inline aloha-h2o/pom.xml:25]: Updated.

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    * Currently (3.6.0.3), h2o generated model says: "" when a categorical value is not supplied.
+    * This is determined from inspecting the generated H2o model code so it's likely brittle and subject
+    * to change but its better than throwing an IllegalArgumentException with no diagnostics information
+    * when there is missing data in a categorical variable.
+    * @param e exception thrown by h2o
+    * @param f the data passed in.
+    * @return whether to attempt to recover.  Don't attempt to recover unless a string-based feature appears to
+    *         be missing.  This is so that we can diagnose when the model will fail every time.
+    */
+  protected[this] def isCategoricalMissing(e: IllegalArgumentException, f: Features[RowData]): Boolean =
```
Comment (deaktator) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala:97]: Look here.  I tried to make this forgiving as to the error message but I wanted to include the check for the feature specification of a string-based feature.  See scaladoc comments above.

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    else false
+
+  /**
+    * Report a problem presumably resulting from a missing categorical variable.
+    * @param t the error to be reported.
+    * @param f the feature values
+    * @param audit whether to audit the score
+    * @return
+    */
+  protected[this] def handleMissingCategorical(t: IllegalArgumentException, f: Features[RowData])(implicit audit: Boolean) = {
```
Comment (deaktator) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/H2oModel.scala:114]: Check here too.

File: aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/categories/package.scala [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    TypeCoercion[C, B] map { coercion =>
+      predictionFn(model) andThen (_.right map coercion)
+    } toRight noCoercionFound(model)
+
+  private[this] def noCoercionFound(model: EasyPredictModelWrapper) =
+    TypeCoercionNotFound(model.getModelCategory)
+}
+
+object H2oModelCategory {
+  private[categories]def wellConditioned(d: Double) = jl.Double.compare(NaN, d) != 0 &&
```
Comment (deaktator) [inline aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/categories/package.scala:35]: Fixed `NaN` issue here.

**Comment by deaktator** [comment]
Updated the POM so that it compiles the submodules in the proper order.

**Comment by deaktator** [comment]
Added CLI and tests
